### PR TITLE
Chulanov, LETI, added the 2nd lab

### PR DESCRIPTION
**Pull request by Dron642**
Chulanov Andrey, leti, 2381
Added 1st lab.

### CODE REVIEW

**Comment by Dron642** [comment]
Added the 2nd lab

File: leti/2381/chulanov/1/main.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+
+int condition = 0;
+bool consumer_starts=false;
+bool continue_updates = true;
+
+ 
+void* producer_routine(void* arg) {
+  // Wait for consumer to start
+    while (!consumer_starts)
```
Comment (eugenyk) [inline leti/2381/chulanov/1/main.cpp]: При сборке с полной оптимизацией и запуске на 4 ядрах данный метод не уходит после запуска из цикла. Всё висит сразу после запуска с таким стеком:

(gdb) bt
#0  0x00007ff887d806bd in pthread_join (threadid=140705403725568, thread_return=0x0) at pthread_join.c:90
#1  0x0000000000401191 in run_threads() ()
#2  0x0000000000400dc9 in main ()
(gdb) info threads
  Id   Target Id         Frame 
  4    Thread 0x7ff8879b6700 (LWP 3076) "a.out" 0x0000000000401019 in producer_routine(void*) ()
  3    Thread 0x7ff8871b5700 (LWP 3077) "a.out" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  2    Thread 0x7ff8869b4700 (LWP 3078) "a.out" 0x0000000000400fd9 in consumer_interruptor_routine(void*) ()
* 1    Thread 0x7ff888a18740 (LWP 3075) "a.out" 0x00007ff887d806bd in pthread_join (threadid=140705403725568, thread_return=0x0) at pthread_join.c:90
(gdb) thread 2
[Switching to thread 2 (Thread 0x7ff8869b4700 (LWP 3078))]
#0  0x0000000000400fd9 in consumer_interruptor_routine(void*) ()
(gdb) bt
#0  0x0000000000400fd9 in consumer_interruptor_routine(void*) ()
#1  0x00007ff887d7f61a in start_thread (arg=0x7ff8869b4700) at pthread_create.c:334
#2  0x00007ff887ab95fd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
(gdb) thread 3
[Switching to thread 3 (Thread 0x7ff8871b5700 (LWP 3077))]
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
185     62:     movl    (%rsp), %edi
(gdb) bt
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
#1  0x0000000000400f87 in consumer_routine(void*) ()
#2  0x00007ff887d7f61a in start_thread (arg=0x7ff8871b5700) at pthread_create.c:334
#3  0x00007ff887ab95fd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
(gdb) thread 4
[Switching to thread 4 (Thread 0x7ff8879b6700 (LWP 3076))]
#0  0x0000000000401019 in producer_routine(void*) ()
(gdb) bt
#0  0x0000000000401019 in producer_routine(void*) ()
#1  0x00007ff887d7f61a in start_thread (arg=0x7ff8879b6700) at pthread_create.c:334
#2  0x00007ff887ab95fd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109

File: leti/2381/chulanov/2/lab2/LockFreeSetImpl.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            
+            if(head.next.getReference() == tail){
+                return true;
+            }
+            // test if all elements are marked to remove
+            else{
+                Node<T> head_next = head.next.get(marked);
+                if(marked[0] && head_next != tail){
+                    Node<T> head_next_next = head_next.next.getReference();
+                    head.next.compareAndSet(head_next,head_next_next,false,false);
```
Comment (eugenyk) [inline leti/2381/chulanov/2/lab2/LockFreeSetImpl.java]: Нужно пояснение: сюда мы приходим, когда marked[0] истинно, а в cas мы предполагаем, что marked = false, нестыковка?

**Comment by Dron642** [comment]
Исправил.
Во второй работе проверял не тот элемент на marked, поэтому несостыковка.

**Pull request merged**